packages/realm/src/IndirectWeakCache.ts

```diff
@@ -0,0 +1,56 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2022 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { HashFunction, IndirectWeakMap } from "./internal";
+
+/**
+ * A cache of objects (the value) which can either be constructed on demand or retrieved from cache.
+ * The cache is considered weak as it extends the `IndirectWeakMap` to store its values, making them available
+ * for garbage collection.
+ * @internal
+ */
+export class IndirectWeakCache<
+  K extends object,
```

Comment:
In my opinion, we should start naming our generics.  I always have to read the implementation to deduce what these letters actually mean.  Thoughts?

Reply:
I agree .. ol' habits stick around.
I pushed a commit with some less abbreviated names 🙂

Reply:
Meh. I think K and V are pretty classic abbreviations for Key and Value. And I don't think the Type suffix adds any value, both because they are capitalized, and because types are the only thing that TS supports as generic arguments (even things that look "valuey" like `true` and `"hello"` are promoted to "literal types" and are still considered types).

I agree that `Args` and `Hash` are good names though. But we don't need to go all Java Enterprise Edition on all of our names.

Reply:
Time to flip a 🪙 ? I don't feel strong either way.

Reply:
I mean, i'm not gonna die on this hill, but i've always felt being more explicit is better.  As long as we don't start using T, K, V for anything else that Type, Key and Value, i'm fine with it.  Just be explicit where it matters.